packages/core/src/utils.js

```diff
@@ -93,6 +93,74 @@ export function getSchemaType(schema) {
   return type;
 }
 
+// detects whether a schema references itself recursively
+export function isCyclic(schema, rootSchema) {
+  const traversed = new Set();
+
+  /**
+   * Helper fn to recurse into children
+   * @param  {Array} children   The target array
+   * @return {Boolean}          Returns whether there was a cyclic event in the children
+   */
+  const checkChildren = children => {
+    return children.reduce((acc, child) => {
+      return Boolean(check(child) | acc);
+    }, false);
+  };
+
+  const check = schema => {
+    // handling of malformed data
+    if (typeof schema === "undefined" || typeof schema !== "object") {
+      return false;
+    }
+
+    if (traversed.has(schema)) {
+      return true;
+    }
+
+    traversed.add(schema);
+
+    if ("$ref" in schema) {
+      const refSchema = findSchemaDefinition(schema.$ref, rootSchema);
+      const result = check(refSchema);
+      traversed.delete(schema); // we need to delete here because siblings can refer to the same definition
+      return result;
+    }
+
+    switch (schema.type) {
```

Comment:
Modeling this strictly against the type kind of worries me because this also needs to handle `oneOf`, `anyOf`, `allOf` cases, as well as `if`, `then`, and `else` in the newer versions of JSON Schema (that OAS now supports with the coming v3.1) that's being added in https://github.com/rjsf-team/react-jsonschema-form/pull/1666.

This is how `json-schema-ref-parser` crawls though a schema looking for a circular ref: https://github.com/APIDevTools/json-schema-ref-parser/blob/bd23ead4b3c3605fc34b8b4f31ca81894e095fc0/lib/dereference.js#L35-L82

If you want to add support for `oneOf`, `anyOf`, and `allOf` here that would probably be good enough for our own cases, but we should go to the drawing board and come up with future-proof solution.

Reply:
Could you help me build out a test for `oneOf`, `anyOf`, and `allOf` with regards to infinite recursion?